### PR TITLE
added support for bidExchange and askExchange

### PR DESCRIPTION
**Pull request by mrwillett**
- Added support for displaying the bid and ask exchanges as provided by tick types 32 and 33 [here](https://interactivebrokers.github.io/tws-api/tick_types.html)
- The resulting string of capital letters _appears_ to follow the exchange codes found [here](https://tlc.thinkorswim.com/center/howToTos/thinkManual/Miscellaneous/Exchange-Codes)
- The goal is to be able to decode the exchanges using reqSmartComponents(), though this commit merely adds the encoded exchange sequences to the ticker objects.

This commit passed 100% of the tests.

I appreciate any feedback.

### CODE REVIEW

File: ib_insync/decoder.py [hunk 1]
```diff
@@ -60,6 +60,8 @@ def __init__(self, wrapper: Wrapper, serverVersion: int):
                 'scannerParameters', [str]),
             20: self.scannerData,
             21: self.tickOptionComputation,
+            32: self.wrap('tickString', [str]),
+            33: self.wrap('tickString', [str]),
```
Comment (erdewit) [inline ib_insync/decoder.py:64]: The numbers 32 and 33 are not message types (they're tick types) and these two lines can be removed.

File: ib_insync/ticker.py [hunk 2]
```diff
@@ -159,11 +161,11 @@ def trades(self) -> "Tickfilter":
 
     def bids(self) -> "Tickfilter":
         """Emit bid ticks."""
-        return Tickfilter((0, 1, 66, 69), self)
+        return Tickfilter((0, 1, 32, 66, 69), self)
```
Comment (erdewit) [inline ib_insync/ticker.py:164]: Can you revert this? It should pass only valid TickData.

File: ib_insync/ticker.py [hunk 3]
```diff
@@ -159,11 +161,11 @@ def trades(self) -> "Tickfilter":
...
 
     def asks(self) -> "Tickfilter":
         """Emit ask ticks."""
-        return Tickfilter((2, 3, 67, 70), self)
+        return Tickfilter((2, 3, 33, 67, 70), self)
```
Comment (erdewit) [inline ib_insync/ticker.py:168]: Please revert this as well.

File: ib_insync/wrapper.py [hunk 4]
```diff
@@ -783,7 +783,11 @@ def tickString(self, reqId: int, tickType: int, value: str):
         if not ticker:
             return
         try:
-            if tickType == 47:
+            if tickType == 32:
+                ticker.bidExchange = value
+            elif tickType == 33:
+                ticker.askExchange = value
+            elif tickType == 47:
```
Comment (erdewit) [inline ib_insync/wrapper.py:790]: Can you add a test for tickType == 84 as well, for setting ticker.lastExchange

**Review by erdewit — CHANGES_REQUESTED** [review]
Thanks for the PR. I will add IB.reqSmartComponents as well to see what it does.

File: ib_insync/ticker.py [hunk 5]
```diff
@@ -46,8 +46,10 @@ class Ticker:
     marketDataType: int = 1
     bid: float = nan
     bidSize: float = nan
+    bidExchange: str = ''
     ask: float = nan
     askSize: float = nan
+    askExchange: str = ''
     last: float = nan
```
[the remark this answers is not in the document]
Comment (mrwillett) [reply ib_insync/ticker.py:53]: Added in new commit

File: ib_insync/wrapper.py (hunk 4 above)
Comment (mrwillett) [reply ib_insync/wrapper.py:792]: added in new commit

**Comment by mrwillett** [comment]
Interestingly, when testing the output of the changes made, I couldn't get lastExchange to work for options contracts. It works for stocks, but I couldn't get any output, even for high liquidity options (near-term at-the-money SPY and SPX contracts). So I dug into it and opened an options chain on TWS and sure enough bid and ask exchanges were visible, but lastExchange was blank. I guess it's just another one of the many idiosyncrasies of IB's API. Here's the script I'm using to test the output of the changes:

```py
from ib_insync import IB, Option, Stock


ib = IB()
ib.connect('127.0.0.1', 7497, 0)

# uncomment one of below:
# contract = Option('SPY', '20220401', 420, 'C', 'SMART')
contract = Stock('AAPL', 'SMART', 'USD')
# contract = Option('SPX', '20220401', 4295, 'C', 'CBOE')
data = ib.reqMktData(contract)

try:
    while True:
        ib.sleep(1)
        print('--------------')
        print('bidEx:', data.bidExchange,
              '\naskEx:', data.askExchange,
              '\nlastEx:', data.lastExchange)
except KeyboardInterrupt:
    ib.disconnect()```

**Review by erdewit — APPROVED** [review]

**Comment by erdewit** [comment]
Thanks for the changes. To decode the capital letters there's a newly added ``reqSmartComponents`` request and a new ``Ticker.bboExchange`` field. 

Here's an example of using it - note that the exchange must be open, which is why a Dutch stock is used now:

```python
contract = Stock('PHIA', 'SMART', 'EUR')
ticker = ib.reqMktData(contract)
while not ticker.bboExchange:
    ib.waitOnUpdate()
components = ib.reqSmartComponents(ticker.bboExchange)
print(components)
```

**Comment by erdewit** [comment]
The output is:

```
[SmartComponent(bitNumber=0, exchange='AEB', exchangeLetter='A'),
 SmartComponent(bitNumber=1, exchange='BVL', exchangeLetter='B'),
 SmartComponent(bitNumber=2, exchange='ENEXT.BE', exchangeLetter='E'),
 SmartComponent(bitNumber=3, exchange='SBF', exchangeLetter='S'),
 SmartComponent(bitNumber=4, exchange='TGATE', exchangeLetter='T')]
```

**Comment by mrwillett** [comment]
Very cool. I appreciate your help with this, Ewald!

**Pull request merged**